### ISSUE TITLE
App starts and is running in Heroku but gives 404 error

### ISSUE THREAD

**Issue by kevinsickles** [bbrookfield/alexa-thermostat#7]
I uploaded the code github code to  Heroku, The app starts and says it is up.   I see the requests from Alexa Skills Test,    all requests are returned with a 404.   

any thoughts on changes that need to be made to run on Heroku?  

State changed from up to starting

2016-08-30T17:48:05.230024+00:00 heroku[web.1]: State changed from starting to up
2016-08-30T17:48:11.312717+00:00 heroku[router]: at=info method=GET path="/" host=quiet-caverns-24299.herokuapp.com request_id=733ee5f5-0a4c-47be-b1c8-e5fc5d9c96ee fwd="96.246.173.232" dyno=web.1 connect=1ms service=22ms status=404 bytes=211

Thank you Kevin

**Comment by bbrookfield** [comment]
Kevin, did you set the private bar in your Heroku set up. You need to set
the private bar to the IP address or domain name you use to access your
thermostat

**Comment by kevinsickles** [comment]
I did set both the THERMOSTAT_URL and PORT Config Vars,   Is that what you mean by private var?

**Comment by kevinsickles** [comment]
Started from scratch again,   same issue,   Here is the build log if that helps?  

-----> Node.js app detected
-----> Creating runtime environment

```
   NPM_CONFIG_LOGLEVEL=error
   NPM_CONFIG_PRODUCTION=true
   NODE_ENV=production
   NODE_MODULES_CACHE=true
```

-----> Installing binaries
       engines.node (package.json):  unspecified
       engines.npm (package.json):   unspecified (use default)

```
   Resolving node version (latest stable) via semver.io...
   Downloading and installing node 5.11.1...
   Using default npm version: 3.8.6
```

-----> Restoring cache
       Skipping cache restore (new runtime signature)
-----> Building dependencies
       Installing node modules (package.json)
       alexa-thermostat@1.0.0 /tmp/build_dfadb309d38deca03400b329beca369f
       ├─┬ alexa-app@2.3.4
       │ ├─┬ alexa-utterances@0.2.1
       │ │ └── js-combinatorics@0.5.2
       │ ├── bluebird@2.11.0
       │ └── numbered@1.0.0
       ├── async@1.5.2
       ├─┬ body-parser@1.15.2
       │ ├── bytes@2.4.0
       │ ├── content-type@1.0.2
       │ ├─┬ debug@2.2.0
       │ │ └── ms@0.7.1
       │ ├── depd@1.1.0
       │ ├─┬ http-errors@1.5.0
       │ │ ├── inherits@2.0.1
       │ │ ├── setprototypeof@1.0.1
       │ │ └── statuses@1.3.0
       │ ├── iconv-lite@0.4.13
       │ ├─┬ on-finished@2.3.0
       │ │ └── ee-first@1.1.1
       │ ├── qs@6.2.0
       │ ├─┬ raw-body@2.1.7
       │ │ └── unpipe@1.0.0
       │ └─┬ type-is@1.6.13
       │   └── media-typer@0.3.0
       ├─┬ express@4.14.0
       │ ├─┬ accepts@1.3.3
       │ │ └── negotiator@0.6.1
       │ ├── array-flatten@1.1.1
       │ ├── content-disposition@0.5.1
       │ ├── cookie@0.3.1
       │ ├── cookie-signature@1.0.6
       │ ├── encodeurl@1.0.1
       │ ├── escape-html@1.0.3
       │ ├── etag@1.7.0
       │ ├── finalhandler@0.5.0
       │ ├── fresh@0.3.0
       │ ├── merge-descriptors@1.0.1
       │ ├── methods@1.1.2
       │ ├── parseurl@1.3.1
       │ ├── path-to-regexp@0.1.7
       │ ├─┬ proxy-addr@1.1.2
       │ │ ├── forwarded@0.1.0
       │ │ └── ipaddr.js@1.1.1
       │ ├── range-parser@1.2.0
       │ ├─┬ send@0.14.1
       │ │ ├── destroy@1.0.4
       │ │ └── mime@1.3.4
       │ ├── serve-static@1.11.1
       │ ├── utils-merge@1.0.0
       │ └── vary@1.1.0
       ├── q@1.4.1
       └─┬ request@2.74.0
       ├── aws-sign2@0.6.0
       ├── aws4@1.4.1
       ├─┬ bl@1.1.2
       │ └─┬ readable-stream@2.0.6
       │   ├── core-util-is@1.0.2
       │   ├── isarray@1.0.0
       │   ├── process-nextick-args@1.0.7
       │   ├── string_decoder@0.10.31
       │   └── util-deprecate@1.0.2
       ├── caseless@0.11.0
       ├─┬ combined-stream@1.0.5
       │ └── delayed-stream@1.0.0
       ├── extend@3.0.0
       ├── forever-agent@0.6.1
       ├─┬ form-data@1.0.1
       │ └─┬ async@2.0.1
       │   └── lodash@4.15.0
       ├─┬ har-validator@2.0.6
       │ ├─┬ chalk@1.1.3
       │ │ ├── ansi-styles@2.2.1
       │ │ ├── escape-string-regexp@1.0.5
       │ │ ├─┬ has-ansi@2.0.0
       │ │ │ └── ansi-regex@2.0.0
       │ │ ├── strip-ansi@3.0.1
       │ │ └── supports-color@2.0.0
       │ ├─┬ commander@2.9.0
       │ │ └── graceful-readlink@1.0.1
       │ ├─┬ is-my-json-valid@2.13.1
       │ │ ├── generate-function@2.0.0
       │ │ ├─┬ generate-object-property@1.2.0
       │ │ │ └── is-property@1.0.2
       │ │ ├── jsonpointer@2.0.0
       │ │ └── xtend@4.0.1
       │ └─┬ pinkie-promise@2.0.1
       │   └── pinkie@2.0.4
       ├─┬ hawk@3.1.3
       │ ├── boom@2.10.1
       │ ├── cryptiles@2.0.5
       │ ├── hoek@2.16.3
       │ └── sntp@1.0.9
       ├─┬ http-signature@1.1.1
       │ ├── assert-plus@0.2.0
       │ ├─┬ jsprim@1.3.0
       │ │ ├── extsprintf@1.0.2
       │ │ ├── json-schema@0.2.2
       │ │ └── verror@1.3.6
       │ └─┬ sshpk@1.10.0
       │   ├── asn1@0.2.3
       │   ├── assert-plus@1.0.0
       │   ├─┬ bcrypt-pbkdf@1.0.0
       │   │ └── tweetnacl@0.14.3
       │   ├─┬ dashdash@1.14.0
       │   │ └── assert-plus@1.0.0
       │   ├── ecc-jsbn@0.1.1
       │   ├─┬ getpass@0.1.6
       │   │ └── assert-plus@1.0.0
       │   ├── jodid25519@1.0.2
       │   ├── jsbn@0.1.0
       │   └── tweetnacl@0.13.3
       ├── is-typedarray@1.0.0
       ├── isstream@0.1.2
       ├── json-stringify-safe@5.0.1
       ├─┬ mime-types@2.1.11
       │ └── mime-db@1.23.0
       ├── node-uuid@1.4.7
       ├── oauth-sign@0.8.2
       ├── stringstream@0.0.5
       ├── tough-cookie@2.3.1
       └── tunnel-agent@0.4.3

-----> Caching build
       Clearing previous node cache
       Saving 2 cacheDirectories (default):
       - node_modules
       - bower_components (nothing to cache)
-----> Build succeeded!
       ├── alexa-app@2.3.4
       ├── async@1.5.2
       ├── body-parser@1.15.2
       ├── express@4.14.0
       ├── q@1.4.1
       └── request@2.74.0

-----> Discovering process types
       Procfile declares types -> web
-----> Compressing...
       Done: 13.7M
-----> Launching...
       Released v3
       https://serene-woodland-90278.herokuapp.com/ deployed to Heroku

**Comment by kevinsickles** [comment]
Also did set the following vars and I can connect to the thermostat using the URL and Port

Config Vars

PORT     9010

THERMOSTAT_URL    kevinsickles.xxxxx.com

**Comment by dvideby0** [comment]
I believe the port should be set by heroku itself.. Do you see any errors
around the process not binding to a port in heroku? if so, set port to be
process.env.PORT (case sensitive)

**Comment by dvideby0** [comment]
I believe the port should be set by heroku itself.. Do you see any errors around the process not binding to a port in heroku? if so, set port to be process.env.PORT (case sensitive)

**Comment by dvideby0** [comment]
don't set a port value in heroku. they auto bind to ports. so if you have that as an environment variable in your app you need to remove it. You stuff will be available through standard http(s) on their side because they have a load balancer in front of your app. Make sense?

**Comment by kevinsickles** [comment]
divideby0   Thank you,   just to clarify,  The app.js code deployed ends with  
  app.listen(process.env.PORT);     I am then setting the PORT var in the Heroku dashboard for config vars,   so you are saying remove that variable?   where do you set the port for the thermostat url?

**Comment by kevinsickles** [comment]
When I remove app.listen(process.env.PORT);   app crashes on start
2016-09-01T21:07:09.839726+00:00 heroku[web.1]: State changed from crashed to starting
2016-09-01T21:07:11.216465+00:00 heroku[web.1]: Starting process with command `node app.js`
2016-09-01T21:07:13.629989+00:00 heroku[web.1]: State changed from starting to crashed
2016-09-01T21:07:13.610504+00:00 heroku[web.1]: Process exited with status 0

**Comment by dvideby0** [comment]
you don't want to remove process.env.PORT you just want to make sure you are not setting a value anywhere else. like in your code you can't have process.env.PORT = 9010; or in your environment variables for heroku you can't set PORT to a value. PORT is a reserved environment variable on heroku. It will auto assign it a value. Make sense?

**Comment by kevinsickles** [comment]
It does,   I am not setting a value anywhere else,   I get   "Cannot GET /"  when I try to open the app.   the log shows at=info method=GET path="/" host=serene-woodland-90278.herokuapp.com request_id=d9e40b28-9d10-4b83-b9c5-549c90708c06 fwd="96.246.173.232" dyno=web.1 connect=1ms service=40ms status=404 bytes=211

**Comment by kevinsickles** [comment]
when I run a test from alexa skills  with the set interaction model,      I get the following in the heroku logs.  

State changed from starting to up
2016-09-01T21:55:43.304005+00:00 app[web.1]: Error: Error: Invalid URI "kevinsickles.xxxxxxx.com/tstat" RESPONSE: undefined BODY: undefined
2016-09-01T21:55:43.311865+00:00 app[web.1]: Unhandled rejection Error: Unhandled exceptionUnexpected token u
2016-09-01T21:55:43.311867+00:00 app[web.1]:     at Object.ensureErrorObject (/app/node_modules/bluebird/js/main/util.js:261:20)
2016-09-01T21:55:43.311868+00:00 app[web.1]:     at /app/node_modules/bluebird/js/main/promise.js:489:17
2016-09-01T21:55:43.311868+00:00 app[web.1]:     at response.fail (/app/node_modules/alexa-app/index.js:248:11)
2016-09-01T21:55:43.311869+00:00 app[web.1]:     at /app/node_modules/alexa-app/index.js:301:20
2016-09-01T21:55:43.311869+00:00 app[web.1]:     at tryCatcher (/app/node_modules/bluebird/js/main/util.js:26:23)
2016-09-01T21:55:43.311870+00:00 app[web.1]:     at new Promise (/app/node_modules/bluebird/js/main/promise.js:71:37)
2016-09-01T21:55:43.311871+00:00 app[web.1]:     at /app/app.js:40:32
2016-09-01T21:55:43.311872+00:00 app[web.1]:     at next (/app/node_modules/express/lib/router/route.js:131:13)
2016-09-01T21:55:43.311872+00:00 app[web.1]:     at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
2016-09-01T21:56:13.266844+00:00 heroku[router]: at=error code=H12 desc="Request timeout" method=POST path="/thermostat" host=stark-garden-42977.herokuapp.com request_id=924f233a-4062-428c-8903-9bef9403acca fwd="72.21.217.78" dyno=web.1 connect=0ms service=30000ms status=503 bytes=0

**Comment by kevinsickles** [comment]
Finally got it,    needed to put http://   infront of the URL var

**Comment by kevinsickles** [comment]
P.S.   For heroku the destination port does need to be set in the URL var    Http://xxx.xxx.xxx.xxx:port   not by using the PORT var  or it defaults to port 80

**Comment by droidforever** [comment]
Okay let me start by saying I have no idea what I am doing.  Complete N00b!  I am having a similar problem.  I uploaded the code to Heroku.  Here are the steps I took:
heroku login
git clone git clone https://github.com/bbrookfield/node-alexa-server
cd node-alexa-server
heroku create my-thermostat
git push heroku master
"Showed Success"
went to my dashboard and clicked on my-thermostat app then went to settings
in Config vars I added
THERMOSTAT_URL https://mydomain.com (also tried https://mydomain.com:5000 and https://mydomain.com/thermostat)
went back to cmd
heroku ps:scale web=1
heroku open
Log Results;
at=info method GET path="/" host=my-thermostat.herokuapp.com request_id=5bb886c9-90cc-4584-8762-d88228919798 fwd="xxx.xxx.xxx.xxx" dyno=web.1 connect=1ms service=22ms status=404 bytes=211
I have done this several times and nothing seems to work. Obviously I am doing something drastically wrong.  Any help is greatly appreciated.

**Comment by kevinsickles** [comment]
The thermostat call should be http   not https.    Have you built the skill in Amazon Voice services?  You want to test the Heroku app from there.

**Comment by droidforever** [comment]
Yes I built the skill in the Alexa skills app and tried to test from there was returned an error the end point could not be reached or something similar.  I have the skills pointed to https://mydomain.com/thermostat and I have tried both http and https in the heroku app and still get the same results.  The skill does show up in my android echo app but will not connect.  Thank you for responding.

**Comment by kevinsickles** [comment]
OK,   just to clarify,   The URL in the Alexa skill is the heroku apps https url?   and the url in the heroku app is the public facing url that can reach your thermostat?

**Comment by droidforever** [comment]
No the URL in Alexa skills kit is https://mydomain.com/thermostat and in heroku is http://mydomain.com:5000 which is their default port but I have tried matching URL's also

**Comment by kevinsickles** [comment]
Heroku generates a URL when you build the app.   That goes in the Alexa
skills endpoint

Thank you

Kevin Sickles
914-315-9483

**Comment by bbrookfield** [comment]
A couple of things to consider

The way that this skill works is as follows

You say a request to alexa
     "Alexa ask MyThermostat what is the temperature"

The echo sends the request to Amazon which then calls the skill "MyThermostat" using the URL you defined as the endpoint for your skill passing the "What is the temperature" as a request

You skill parses the request and send the TSTAT command to the IP Address you defined for your skill in Huroku config Variable THERMOSTAT_URL

Your Router receives then forwards the request "TSTAT" to the IP address of your thermostate

Your thermostat creates the response that is sent back to you Skill on Huroku, which then sends back the response to Amazon

Amazon then send the response back to your echo.

 

1)  The Alexa skill must be called by Amazon Using HTTPS. This is why I
suggest Heroku. It provides a free cloud server that has a valid HTTPS
certificate.

The URL to set up in AMAZON will be defined by Huroku and will look
something like

https://yourskillname.herokuapp.com/thermostat


2)  The URL you code in the SETTINGS -> Config Variables for the thermostat Alexa app is
the url to you home router. This URL must start with HTTP or HTTPS (If you
have a SSL certificate for you domain)

3)  The end of the URL should have a port number that you router will use
to forward the request to your thermostat

Finding the IP address of your thermostat must be done before you can setup
the port forward in your router

4)  You have to setup a specific Port forwarding so that the request
received by your router from the Alexa app on Heroku is sent to the IP
address of your thermostat

Setting up a port forward varies from router to router, but the general
idea is to pick a port number that does not conflict with  any other known
IP services( I used port 9013). The URL you setup in Heroku settings for the thermostat app would look
like

http://mydomainname.com:9013

first test that you can talk to the thermostat using a your web browser (Same network)

example: 192.168.1.56/tstat (Assuming that 192.168.1.56 is ip address of your thermostat)

You should get back something like

{"temp":75.00,"tmode":3,"fmode":0,"override":0,"hold":0,"
tstate":0,"fstate":0,"time":{"day":5,"hour":23,"minute":54},"t_type_post":0}

This confirms that you  have the correct IP address of your thermostat

Next test by using a browser not on you home network (I use my Cell phone )
and use url

http://mydomainname.com:9013/tstat

This will confirm that you can talk to your thermostat outside of your home
network and that the thermostat app on Heroku can send requests

you should get back same result

If not then you router port forward is not working.

If it works, then the Thermostat skill should work. If not then check/Verify the endpoint setting in the Alexa Dev console and try testing your skill using the TEST option in the Alexa Dev console.


Hope this helps.


Bill B

**Comment by droidforever** [comment]
okay I have done as suggested above.  I am still getting Cannot GET /  when I try to open the app from Heroku.

 When I try to browse to https://mydomain.com:7575/tstat the address fails to connect.

I have opened port 7575 both UDP and TCP to my thermostat IP and I do get a response from http://my_internal_IP:7575/tstat

In heroku I have Config Vars as THERMOSTAT_URL set to http://mydomainname.com:7575 

The URL in the heroku app is listed as https://my_skill_name.herokuapp.com 

I have that in my Amazon skills config and after entering I have tried both I have a signed root certificate and am a subdomain of a signed certificate. 
You guys have been great with your assistance I truly appreciate all you have done.  So sorry for being such a n00b!

**Comment by kevinsickles** [comment]
When I try to browse to https://mydomain.com:7575/tstat the address fails to connect.

Unless you have ssl on your home router you need to usr http.  Does it work for http://mydomain.com:7575/tstat?    If not, port forwarding is not set correctly

**Comment by droidforever** [comment]
No neither one works.  I do have a secure certificate for my domain from Let's Encrypt.  You say port forwarding is configured improperly but I have both tcp/udp connections open to the IP of my thermostat.  Do I need to setup some sort of proxy?

**Comment by bbrookfield** [comment]
your port forward should be setup as

external port = 7575
internal port = 80
ip address = <internal Ip address of your thermostat>

You said you get a response from http://my_internal_ip:7575/tstat
this should not work if "my_internal_ip" is the ip address of your
thermostat as it is expected the request to come in on port 80

Bill B

**Comment by droidforever** [comment]
thank you I addressed this wrong.  192.168.0.120/tstat gives response https://mydomain.com:7575/tstat does not.  Port 80 is opened to my web server port 7575 is opened to my thermostat. So http://mydomain.com and https://mydomain.com bring up an index.html page.  So if I understand this correctly I will need a proxy to pass my web server to something like port 81 so I can use port 80 internally for the thermostat?  I am using IIS if that makes a difference.

**Comment by bbrookfield** [comment]
You do not need a proxy, What you need is to setup a PORT FORWARD on your
router.

A port forward setup will redirect incoming requests from the internet on
port 7575 to the internal ip address of your thermostat on port 80.
Only requests on port 7575 will be routed.
requests coming in on port 80 will go to your web server just as before


What kind of router do you have?  You can google how to setup a port
forward for your specific router.
If you send me your phone number, I can try and talk you thru it

Bill B

**Comment by droidforever** [comment]
I think I understand now.
Port forward port 80 to the webserver both UDP/TCP on server IP of 192.168.0.200
Port forward port 7575 to the thermostat UDP only on thermostat IP of 192.168.0.120
My router is ASUS RT-AC87R with DD-WRT firmware
here is a pic of my port forwarding
https://jdmguide.com/pic.html

**Comment by bbrookfield** [comment]
Yes that is correct but you don't have to port forward to your web server
since the incoming Port is 80 for normal HTTP request that don't have a
port on the URL line since Port 80 is default port for http for https the
port is 443.

I have used DDWRT before and I know it has port forwarding as a setup option

You are on the right track

Bill B

On Jan 23, 2017 9:19 PM, "droidforever" <notifications@github.com> wrote:

I think I understand now.
Port forward port 80 to the webserver both UDP/TCP on server IP of
192.168.0.200
Port forward port 7575 to the thermostat UDP only on thermostat IP of
192.168.0.120
My router is ASUS RT-AC87R with DD-WRT firmware

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/bbrookfield/node-alexa-server/issues/7#issuecomment-274681968>,
or mute the thread

**Comment by droidforever** [comment]
Thanks for the words of encouragement but I've had it set that way since the start of the project.  So I guess there must be something else wrong.  Again thank you both for all your assistance on this

**Comment by kevinsickles** [comment]
Just to clarify the port forwarding for the thermostat in DDwrt.    The port from should be 7575, the ip address is internal address of thermostat    the port to is 80 

https://www.dd-wrt.com/wiki/index.php/Image:TriggeredPortForward.jpg

**Comment by droidforever** [comment]
So you're saying instead of using port forwarding I should use port triggering.  Opening port 5000 which is the heroku server port and then setting a trigger to 80 and 7575?

**Comment by kevinsickles** [comment]
No,   If the Heroku app is calling   http://mydomain.com:7575/tstat your router has to forward port 7575 to port 80.

As Bill has detailed,   until you get a response from http://mydomain.com:7575/tstat  you do not have port forwarding correctly configured on your router.

**Comment by droidforever** [comment]
I am really starting to get frustrated.  I have been developing web sites and running mIRC channels for years.  I understand port forwarding.  I have ports forwarded for other applications such as home assistant but have never used heroku.
I have a SSL cert for my website
I have Kodi running on port 8080 with lambda using python
I have home assistant running on port 8080 using lambda
I have port 7575 forwarded to my thermostat's IP
I have configured everything according to the guide posted above
If I use http://myip/tstat I get a response from the thermostat
{"temp":72.00,"tmode":1,"fmode":0,"override":0,"hold":1,"t_heat":72.00,"tstate":0,"fstate":0,"time":{"day":1,"hour":16,"minute":35},"t_type_post":0}
If I type http://mydomain.com:7575/tstat I get nothing
If I do a netstat -an it does not show port 7575 open which means there is nothing running on that port
port 80 is listed because my web-server is running
port 8123 is listed because hoem assistant is running
port 7575 is not listed as there is no service running on port 7575
Is there some sort of app I need to have running locally?
Do I need to run herloku local? or heroku local web? so that the actual app is running on my server?
Something is missing in my configuration and I can't figure out what it is.

**Comment by kevinsickles** [comment]
You keep saying you have 7575 forward to your thermostat, but you have to forward Port 7575 to Port 80,    on your home router.   for your thermostat's internal ip,

**Comment by kevinsickles** [comment]
Service Name	 Port Range	Local IP	              Local Port	      Protocol	
thermostat	  63999	        192.168.15.75	      80	               TCP

**Comment by bbrookfield** [comment]
You state then
"I have port 7575 forwarded to my thermostat's IP"
do you have the destination port set to 80?

Can you screen shot you portforwarding setup for 7575 on your router so we
can see it?

Doing a netstat on your computer would show nothing as the request from the
router would NOT go to your computer so port 7575 would not show as open
The request should go to the IP address of your thermostat assuming the
port forwarding setup is correct.

Bill B

**Comment by droidforever** [comment]
I fixed the port forward and now when I go to http://mydomain.com:7575/tstat the thermostat responds but I still get Cannot GET / from the heroku url and I get an endpoint error when testing inside the alexa skill do I need to redo the app now that the port is configured properly?

**Comment by kevinsickles** [comment]
Hopefully not,   In the heroku app,   got  to settings  then click on config vars,   put in just one config var.       THERMOSTAT_URL         http://mydomain.com:7575

**Comment by droidforever** [comment]
Thank you but I have already done that.

**Comment by bbrookfield** [comment]
OK,

Now that you can get a response from http://mydomain.com:7575/tstat the next step is
to verify you Huroku and Amazon Setup.

Assuming a name of "MyHurokuApp" is the name of your huroku app 
the endpoint defined in amazon for you skill should be
https://MyHurokuApp.herokuapp.com/thermostat

Bill B

**Comment by bbrookfield** [comment]
You cannot use browser to talk to the Heroku app. It is designed to only be invoked by Amazon requests coming from your echo or from the TEST option in the developer portal for the skill

**Comment by droidforever** [comment]
thank you gentlemen.  My appreciation is abundant.  Everything is working.  I'm so sorry for all the problems this should have been easy.  My only excuse is that I'm old lol.

**Comment by bbrookfield** [comment]
No problem. Glad you got it all working.  I think I will update the readme document based on some of the issues you and others have dealt with.

**Comment by droidforever** [comment]
Bill I am not having great success with heroku due to it timing out I believe maybe a time out issue. So I would like to try this deploying your code locally.  I have a problem
When I type $PORT=xxxx node app.js
I get 'PORT' is not recognized as an internal or external command, operable program or batch file.
Could you tell me how to fix this please.  Sorry to bother you again.

**Comment by kevinsickles** [comment]
I have been using heroku for 6 months,   occasionally it times out and I have to repeat the command.    it the always works on the 2nd command.    For some reason the time outs have become very less common over time.

**Comment by droidforever** [comment]
Here is what I am having problems with.  The first couple commands I test in the developer portal give a response.  Then the responses stop until I run these two commands heroku ps:scale web=0 then heroku ps:scale web=1 and try the commands again and I get a couple more responses.  When the responses are not working I get endpoint can not be reached. I did have to edit the intents and add getHeat and getCool for if I reduced the thermostat the a/c came on if I turned it up my heat came back on. But they both work a couple times and then again I get the endpoint can not be reached.
I really would like to get this working as a local node.js server rather than using a third party like heroku.  But I can't get past the PORT=xxxx  node app.js problem I posted above so therefore I am unable to set a port.  I can manually add the port into the app.js but that doesn't seem to work either.

**Comment by droidforever** [comment]
Here is my log from heroku logs --tail
2017-01-26T07:52:17.312982+00:00 heroku[web.1]: Starting process with command `node app.js`
2017-01-26T07:52:21.035302+00:00 heroku[web.1]: State changed from starting to up
2017-01-26T07:52:56.521124+00:00 heroku[router]: at=error code=H12 desc="Request timeout" method=POST path="/thermostat" host=hatter-thermostat.herokuapp.com request_id=f16c6875-dabc-4b83-bde0-abfb6d90a44e fwd="72.21.217.174" dyno=web.1 connect=0ms service=30000ms status=503 bytes=0
2017-01-26T07:54:33.777230+00:00 app[web.1]: Error: Error: connect ETIMEDOUT 192.168.0.120:80 RESPONSE: undefined BODY: undefined
2017-01-26T07:54:33.779632+00:00 app[web.1]: undefined:1
2017-01-26T07:54:33.779634+00:00 app[web.1]: undefined
2017-01-26T07:54:33.779635+00:00 app[web.1]: ^
2017-01-26T07:54:33.779636+00:00 app[web.1]:
2017-01-26T07:54:33.779637+00:00 app[web.1]: SyntaxError: Unexpected token u in JSON at position 0
2017-01-26T07:54:33.779638+00:00 app[web.1]:     at Object.parse (native)
2017-01-26T07:54:33.779638+00:00 app[web.1]:     at Request._callback (/app/app_modules/alexa-radiothermostat/index.js:55:17)
2017-01-26T07:54:33.779639+00:00 app[web.1]:     at self.callback (/app/node_modules/request/request.js:186:22)
2017-01-26T07:54:33.779640+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-01-26T07:54:33.779641+00:00 app[web.1]:     at Request.emit (events.js:188:7)
2017-01-26T07:54:33.779641+00:00 app[web.1]:     at Request.onRequestError (/app/node_modules/request/request.js:845:8)
2017-01-26T07:54:33.779642+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-01-26T07:54:33.779642+00:00 app[web.1]:     at ClientRequest.emit (events.js:188:7)
2017-01-26T07:54:33.779643+00:00 app[web.1]:     at Socket.socketErrorListener (_http_client.js:310:9)
2017-01-26T07:54:33.779644+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-01-26T07:54:33.779644+00:00 app[web.1]:     at Socket.emit (events.js:188:7)
2017-01-26T07:54:33.779645+00:00 app[web.1]:     at emitErrorNT (net.js:1276:8)
2017-01-26T07:54:33.779645+00:00 app[web.1]:     at _combinedTickCallback (internal/process/next_tick.js:74:11)
2017-01-26T07:54:33.779646+00:00 app[web.1]:     at process._tickCallback (internal/process/next_tick.js:98:9)
2017-01-26T07:54:33.855341+00:00 heroku[web.1]: Process exited with status 1
2017-01-26T07:54:33.896211+00:00 heroku[web.1]: State changed from up to crashed``

**Comment by jswim788** [comment]
If I port forward the thermostat IP, can't anyone on the Internet control my thermostat?  I don't think there is any security built into it.  How are you handling this security?  I think I'm missing a piece of the puzzle.  thanks.
